crates/red_knot_module_resolver/src/resolver.rs

```diff
@@ -143,67 +155,300 @@ impl RawModuleResolutionSettings {
-            }),
-        );
+        static_search_paths.insert(Arc::new(
+            ModuleResolutionPathBuf::first_party(SystemPath::absolute(
+                workspace_root,
+                current_directory,
+            ))
+            .unwrap(),
+        ));
+
+        static_search_paths.insert(Arc::new(custom_typeshed.map_or_else(
+            ModuleResolutionPathBuf::vendored_stdlib,
+            |custom| {
+                ModuleResolutionPathBuf::stdlib_from_custom_typeshed_root(&SystemPath::absolute(
+                    custom,
+                    current_directory,
+                ))
+                .unwrap()
+            },
+        )));
+
+        let mut site_packages = None;
+
+        if let Some(path) = site_packages_setting {
+            let site_packages_root = Arc::new(
+                ModuleResolutionPathBuf::site_packages(SystemPath::absolute(
+                    path,
+                    current_directory,
+                ))
+                .unwrap(),
+            );
+            site_packages = Some(site_packages_root.clone());
+            static_search_paths.insert(site_packages_root);
+        }
 
         // TODO vendor typeshed's third-party stubs as well as the stdlib and fallback to them as a final step
-        if let Some(site_packages) = site_packages {
-            paths.push(ModuleResolutionPathBuf::site_packages(site_packages).unwrap());
-        }
 
         ModuleResolutionSettings {
             target_version,
-            search_paths: OrderedSearchPaths(paths.into_iter().map(Arc::new).collect()),
+            search_path_settings: ValidatedSearchPathSettings {
+                static_search_paths,
+                site_packages,
+            },
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+struct ValidatedSearchPathSettings {
+    /// Search paths that have been statically determined purely from reading Ruff's configuration settings.
+    /// These shouldn't ever change unless the config settings themselves change.
+    ///
+    /// We use an `OrderSet` for storing the paths, since the order is important,
+    /// but at runtime, [`sys.path` never contains duplicate entries]
+    ///
+    /// Note that `site-packages` *is included* as a search path in this sequence,
+    /// but it is also stored separately so that we're able to find editable installs later.
+    ///
+    /// [`sys.path` never contains duplicate entries]: https://docs.python.org/3/library/site.html#module-site
+    static_search_paths: OrderSet<SearchPathRoot>,
+    site_packages: Option<SearchPathRoot>,
+}
+
+/// Collect all dynamic search paths:
+/// search paths listed in `.pth` files in the `site-packages` directory
+/// due to editable installations of third-party packages.
+#[salsa::tracked(return_ref)]
+pub(crate) fn dynamic_module_resolution_paths(db: &dyn Db) -> OrderSet<SearchPathRoot> {
+    db.report_untracked_read();
```

Comment:
Nit: Should these method be called `editable_install_resolution_paths` to better align with the naming of the path variant?

Reply:
I guess I wanted to have a clean distinction between "static" resolution paths that can be derived solely from the Ruff settings given to us by the user, and "dynamic" paths that we have to figure out for ourselves by iterating through `.pth` files. And I suppose there's a possibility that we might want to add support for some other kind of "dynamic" resolution paths in the future (not editables), which we might also want to implement as part of this routine.

But I'm okay with this proposed rename for now; that's pretty hypothetical, and I don't have any specific idea of what other dynamic search paths there could be.